Release v3.0.0

echarrod opened the pull request:
_Proposed change notes below. There are some other changes, like moving items from the main package to subpackages. The relevant ones are mentioned, but clients shouldn't be using the other ones, so I have omitted them._


We've taken this opportunity to address a number of breaking changes. 
To get these updates, and other ones going forwards, you'll need to update references from `github.com/getyoti/yoti-go-sdk/v2` to `github.com/getyoti/yoti-go-sdk/v3`

### Added
- Error returned can be wrapped, and can be signified as being "temporary", i.e. it might be worth retrying the request (see [README](https://github.com/getyoti/yoti-go-sdk#handling-errors) for example)
### Changed
- Preferred way of initialising the yoti client:
  - `client :=  yoti.Client{sdkId, key}` -> `client, err := yoti.NewClient(sdkId, key)`
- `GetActivityDetails(yotiOneTimeUseToken)` now returns a single error, instead of "error strings". This can wrap multiple errors
- `anchor.value` now returns a single string, rather than `[]string`
- Subpackage creation (moved from core `yoti` package:
  - Aml structs to `aml` subpackage:
    - yoti.AmlAddress -> aml.Address
    - yoti.AmlProfile -> aml.Profile
  - Dynamic Scenario structs moved to `dynamic` subpackage:
    - `&yoti.DynamicScenarioBuilder{}` -> `&dynamic.ScenarioBuilder{}`
  - `share.ExtraData` -> `extra.Data`
  - Share URL creation changed:
    - `yoti.CreateShareURL(&client, &scenario)` -> `client.CreateShareURL(&scenario)`
  - `yoti/attribute` -> `yoti/profile/attribute`
  - `yoti.AgeVerification` -> (yoti/profile/)`attribute.AgeVerification`
- Image.Data is now a function, rather than a field:
  - `selfie.Value().Data` -> `selfie.Value.Data()`
- The following methods can now potentially return an error:
  - `SourceConstraintBuilder.Build()`
  - `WantedAttributeBuilder.Build()`
  - `ThirdPartyAttributeExtensionBuilder.Build()`
- `(&ThirdPartyAttributeExtensionBuilder{}).New().` -> `(&ThirdPartyAttributeExtensionBuilder{})`
### Removed
- yoti.Image (use `media.PNGImage` or `media.JPEGImage` instead)
- `yoti.UserProfile` (use `profile.UserProfile`) instead

davidgrayston commented on requests/client.go:
```diff
@@ -0,0 +1,17 @@
+	Do(*http.Request) (*http.Response, error)
+}
+
+// ClientInterface defines the interface required to Mock the YotiClient for
+// testing
+type ClientInterface interface {
+	MakeRequest(string, string, []byte, bool, ...map[int]string) (string, error)
+	GetSdkID() string
+}
```
I think we can remove this now
```suggestion
```

echarrod replied:
Removed in 93cca17 👍

davidgrayston commented:
I think this point in the release notes should be changed from:
> ### Removed
> - yoti.UserProfile (use yoti.Profile) instead

to:

> ### Changed
> - `yoti.Profile` -> `profile.UserProfile`

Also, the _README_ needs to be updated to reflect this - perhaps we should split the _README_ like other SDKs before release? I also noticed a minor typo in the _README_: "Retrieveing the user profile", which could be corrected whilst doing this.

echarrod commented:
> I think this point in the release notes should be changed from:
> 
> > ### Removed
> > 
> > * yoti.UserProfile (use yoti.Profile) instead
> 
> to:
> 
> > ### Changed
> > 
> > * `yoti.Profile` -> `profile.UserProfile`
> 
> Also, the _README_ needs to be updated to reflect this - perhaps we should split the _README_ like other SDKs before release? I also noticed a minor typo in the _README_: "Retrieveing the user profile", which could be corrected whilst doing this.

@davidgrayston I've split out the README, corrected the typo, updated the snippet. As discussed, have left the release notes as they are, since we "deprecated" `yoti.UserProfile` in a previous release

davidgrayston reviewed (approved):
👍

echarrod commented on dynamic/policy_builder.go:
```diff
@@ -0,0 +1,237 @@
+)
+
+const (
+	authTypeSelfieConst = iota + 1
+	authTypePinConst
```
I agree that does make it more explicit, but it does look like best practice seems to lean towards using `iota` - https://yourbasic.org/golang/iota/

echarrod replied:
As discussed, using integer values adds more clarity, and prevents potential problems in the future

echarrod commented on dynamic/policy_builder.go:
```diff
@@ -0,0 +1,237 @@
+
+// WithWantedAttributeByName adds an attribute by its name. This is not the preferred
+// way of adding an attribute - instead use the other methods below
+func (b *PolicyBuilder) WithWantedAttributeByName(name string, options ...interface{}) *PolicyBuilder {
```
We do have:
> This is not the preferred way of adding an attribute - instead use the other methods below

on this method. But yes, the problem is still present for the functions which call this one. Aside from Examples, how do you think we can improve the documentation around this? I guess it's done this was to future-proof us if we introduce other options

davidgrayston replied:
Ruby documents options like this: https://www.rubydoc.info/gems/yoti/1.8.0/Yoti/DynamicSharingService/DynamicPolicyBuilder#with_date_of_birth-instance_method

I suppose we could just add a short explanation to indicate the allowed types, something like this:
```
// WithDateOfBirth adds the date of birth attribute.
// Options accepts multiple SourceConstraint
```

This would need to be duplicated for each method (and updated when new options are introduced), but at least it helps integrators know what the options are.

echarrod replied:
Added in 24a5056 👍

LAFYoti commented:
Since the AML documentation has been removed shall we also remove the AML example?
